### PR TITLE
docs: add flag config adr

### PR DESCRIPTION
**Pull request by toddbaert**
Another "retrospective" ADR, like the cucumber one.

This explains our JSONLogic choice, shared `$evaluators`, as well as the the semantics of flagd's use of OpenFeature's [resolution reasons](https://openfeature.dev/specification/types#resolution-reason).

### CODE REVIEW

File: docs/architecture-decisions/flag-configuration.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+Feature flag systems require a flexible, safe, and portable way to express targeting rules that can evaluate contextual data to determine which variant of a feature to serve.
+
+flagd's targeting system was designed with several key requirements:
+
+## Requirements
+
+- **Language agnostic**: Rules must be portable across different programming languages, ideally relying on existing expression language(s)
+- **Safe evaluation**: No arbitrary code execution or system access
+- **Deterministic**: Same inputs must always produce same outputs
```
Comment (tangenti) [inline docs/architecture-decisions/flag-configuration.md:20]: $flagd.timestamp seems to break this requirement?

File: docs/architecture-decisions/flag-configuration.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+### JSON Logic as the Foundation
+
+flagd chose **JSON Logic** as its core evaluation engine, implementing a modified version with custom extensions.
+This provides a secure, portable foundation where rules are expressed as JSON objects with operators as keys and parameters as values.
+
+#### Benefits realized
+
+- Rules can be stored in databases, transmitted over networks, shared between frontend/backend, and embedded in Kubernetes custom resources
+- No eval() or code injection risks - computations are deterministic and sand-boxed
```
Comment (tangenti) [inline docs/architecture-decisions/flag-configuration.md:34]: I think this depends on the implementation of the eval engine, and less about the language itself.
For example, a JavaScript JSON logic engine had a CVE that allows command injection: https://nvd.nist.gov/vuln/detail/CVE-2021-4329

File: docs/architecture-decisions/flag-configuration.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+# Flag and Targeting Configuration
+
+## Background
+
+Feature flag systems require a flexible, safe, and portable way to express targeting rules that can evaluate contextual data to determine which variant of a feature to serve.
+
+flagd's targeting system was designed with several key requirements:
+
+## Requirements
```
Comment (tangenti) [inline docs/architecture-decisions/flag-configuration.md:16]: What was the requirement for the typing system back then?  Is weakly typed a requirement or a result?

File: docs/architecture-decisions/flag-configuration.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+These enable:
+
+- IDE support with autocomplete
+- Run-time and build-time validation
+- Separate validation of rules and overall configuration if desired
+
+## Considered Options
+
+- **Custom DSL**: Would require parsers in every language
+- **JavaScript/Lua evaluation**: Security risks and language lock-in
```
Comment (tangenti) [inline docs/architecture-decisions/flag-configuration.md:135]: What was the biggest concern for Lua? 
Lua engine normally comes with a sandboxed VM, which is more lightweight and secure IMO.

File: docs/architecture-decisions/flag-configuration.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+- **CEL**: limited number of implementations at time of decision, can't be directly parsed/validated when embedded in Kubernetes resources
+
+## Consequences
+
+### Positive
+
+- Good, because implementations exist across languages
```
Comment (tangenti) [inline docs/architecture-decisions/flag-configuration.md:142]: True, but also there's not much consistency guarantee across languages.

File: docs/architecture-decisions/flag-configuration.md (hunk 3 above)
Comment (tangenti) [inline docs/architecture-decisions/flag-configuration.md:16]: Mostly due to the JS typing system, JSONLogic's behavior on `undefined` values (context not provided, for example) can be very confusing. Was this considered and evaluated?

File: docs/architecture-decisions/flag-configuration.md (hunk 1 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:20]: I guess it depends on whether or not you consider the timestamp an sort of implicit input... We could change it to "predictable" though.

Comment (tangenti) [reply docs/architecture-decisions/flag-configuration.md:20]: To me, what users provide through the evaluation API are the input, and timestamp is not there.

File: docs/architecture-decisions/flag-configuration.md (hunk 2 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:34]: Ya, that's fair. At least at the time, we saw JSONLogic as _less likely_ to have these sort of issues.

File: docs/architecture-decisions/flag-configuration.md (hunk 3 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:16]: Definitely a result.

File: docs/architecture-decisions/flag-configuration.md (hunk 4 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:135]: Lua was something I briefly advocated for, I've had some experience with it doing nginx customization. I remember a community member being opposed to it on the basis of security concerns, and at the time it seemed like a higher bar.

File: docs/architecture-decisions/flag-configuration.md (hunk 5 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:142]: Well, there's enough at least that our test suite works in Go, Python, Java, JS, and .NET. It covers every operation and a fair few edge cases: https://github.com/open-feature/flagd-testbed/tree/main/gherkin

I acknowledge there's inconsistencies, particularly around `undefined`s. It's one of the reasons the [JSONLogic](https://github.com/json-logic/) org exists (though admittedly it's been hard to make progress with the few volunteers there). I'm sure we could use your help if you're interested in contributing to that.

File: docs/architecture-decisions/flag-configuration.md (hunk 3 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:16]: I can't recall if that particular behavior was discussed.

**Comment by toddbaert** [comment]
@tangenti Thanks for the feedback.

I'd like to make it clear that personally, though I think our current JSONLogic implementation(s) are "good enough", I think there's ways to improve, and alternatives. To name a few:

- Leverage WASM and use a single JSON logic implementation, distributed as a module, for maximum implementation coherence
- Work on improving JSONLogic via https://github.com/json-logic/
- Support additional targeting DSLs (Lua, CEL, others), either via WASM or natively.

I would be open to any of the above, and lend my support, if there's engineering effort to spare for these.

File: docs/architecture-decisions/flag-configuration.md (hunk 1 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:20]: OK I'll change that to "predictable".

File: docs/architecture-decisions/flag-configuration.md (hunk 5 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:142]: My first alternative in [this comment](https://github.com/open-feature/flagd/pull/1637#issuecomment-2949513068) might be a reasonable solution to some implementation inconsistencies. Interested in your take on that; we could do it progressively and it would be non-breaking.

File: docs/architecture-decisions/flag-configuration.md (hunk 1 above)
Comment (toddbaert) [reply docs/architecture-decisions/flag-configuration.md:20]: Actually I think I'll keep `Deterministic` (because I believe it still is), but mention the timestamp exception situation.

**Comment by toddbaert** [comment]
Since this is historical, I'm merging this.

That doesn't mean it can't be edited or superseded, it's simply an explanation of a previous decision and reflects the current implementation.
